src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -66,17 +86,22 @@ internal partial class RequestExecutionQueue
                 string? clientName,
                 string methodName,
                 TextDocumentIdentifier? textDocument,
+                TRequestType request,
+                IRequestHandler<TRequestType, TResponseType> handler,
                 Guid activityId,
                 ILspLogger logger,
                 RequestTelemetryLogger telemetryLogger,
-                Action<Exception> handleQueueFailure,
-                Func<RequestContext?, CancellationToken, Task> callbackAsync,
                 CancellationToken cancellationToken)
             {
+                _completionSource = new TaskCompletionSource<TResponseType?>();
+                // Set the tcs state to cancelled if the token gets cancelled outside of our callback (for example the server shutting down).
+                cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
```

Comment:
@CyrusNajmabadi this was a bit I wasn't sure if it was correct.  But basically if an exception happens before we actually start executing the work (e.g. an exception in creating the request context), then we cancel the queue token and shutdown the server.

However, the test would still be waiting on the result of the TCS which never transitions to the canceled state since the `CallbackAsync` method never executes.  So I register an action on cancellation of the combined token to actually cancel the TCS.